pkg/apis/eventing/v1/trigger_validation.go

```diff
@@ -163,3 +146,72 @@ func (t *Trigger) validateInjectionAnnotation(injectionAnnotation string) *apis.
 	}
 	return nil
 }
+
+func ValidateAttributeFilters(filter *TriggerFilter) (errs *apis.FieldError) {
+	if filter == nil {
+		return nil
+	}
+	return errs.Also(ValidateAttributesNames(filter.Attributes).ViaField("attributes"))
+}
+
+func ValidateAttributesNames(attrs map[string]string) (errs *apis.FieldError) {
+	if attrs == nil {
+		return nil
+	}
+
+	for attr := range attrs {
+		if !validAttributeName.MatchString(attr) {
+			errs = errs.Also(apis.ErrInvalidKeyName(attr, apis.CurrentField, "Attribute name must start with a letter and can only contain lowercase alphanumeric").ViaKey(attr))
+		}
+	}
+	return errs
+}
+
+func ValidateSingleAttributeMap(expr map[string]string) (errs *apis.FieldError) {
+	if len(expr) == 0 {
+		return nil
+	}
+
+	if len(expr) != 1 {
+		return apis.ErrGeneric("Multiple items found, can have only one key-value", apis.CurrentField)
+	}
+	for attr := range expr {
+		if !validAttributeName.MatchString(attr) {
+			errs = errs.Also(apis.ErrInvalidKeyName(attr, apis.CurrentField, "Attribute name must start with a letter and can only contain lowercase alphanumeric").ViaKey(attr))
+		}
+	}
+	return errs
+}
+
+func ValidateSubscriptionAPIFiltersList(filters []SubscriptionsAPIFilter, field string) (errs *apis.FieldError) {
+	if filters == nil {
+		return nil
+	}
+	if len(filters) == 0 {
+		return apis.ErrGeneric(fmt.Sprintf("%s must contain at least one nested filter", field), apis.CurrentField)
+	}
+
+	for i, f := range filters {
+		f := f
+		errs = errs.Also(ValidateSubscriptionAPIFilter(&f)).ViaIndex(i)
```

Comment:
A shallow copy is made for the sole purpose of passing it by reference. We might as well just pass the shallow copy directly to `ValidateSubscriptionAPIFilter`.